test/counter_server_example/src/counter_server_example.erl

```diff
@@ -0,0 +1,261 @@
+%%% @doc
+%%%
+%%% Example for using vnet to simulate a cluster of nodes, one of
+%%% which serves as a counter server.
+%%%
+%%% All _test functions demonstrate scenarios of using the server. See
+%%% their documentation for more info.
+%%%
+%%% All _no_vnet_test functions do not use vnet and their purpose is
+%%% to check the core implementation of the counter server.
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+
+-module(counter_server_example).
+
+%%%-------------------------------------------------------------------
+
+-include_lib("eunit/include/eunit.hrl").
+
+-concuerror_options(
+   [ {instant_delivery, false}
+   , assertions_only
+   , {ignore_error, deadlock}
+   ]).
+
+%%%-------------------------------------------------------------------
+
+-export(
+   [ setup_server/0
+   , once_valid_client/0
+   , twice_valid_client/0
+   , invalid_client/0
+   , twice_valid_same_gen_client/0
+   ]).
+
+%%%-------------------------------------------------------------------
+
+-include("counter_server.hrl").
+
+%%%-------------------------------------------------------------------
+
+%% @doc
+%% Simulating a scenario with one client, making requests and checking
+%% results, without using any vnet logic.
+valid_no_vnet_test() ->
+  {ok, P1} = counter_server_supervisor:start_link_no_vnet(),
+  %% Unlinking to not send signals to supervisor from main process
+  %% exiting.
+  unlink(P1),
+  valid_client_no_vnet().
+
+%%%-------------------------------------------------------------------
+
+%% @doc Client making a second request if the first was successful and
+%% verifying that the second answer is greater than the first.
+valid_client_no_vnet() ->
+  Res1 = valid_local_request(),
+  case Res1 =:= error of
+    true -> ok;
```

Comment:
There's a `check_results/2` branch for handling `Res1 =:= error`, so why don't you do a second request here?

Reply:
This is purely for the sake of reducing the number of schedulings that cannot expose errors. If the first request has failed there is nothing to compare against.